Miscellaneous compile time and run time fixes

WonderfulVoid opened the pull request:
Miscellaneous fixes that were required when porting OFP to a bare metal environment with a custom ODP implementation.
Commit c0f98a8 should be of special interest since I think without this, all port hash table computations will collapse into slot 0 of the hash table (the uninitialized port hash mask is probably still 0) which is detrimental to performance.

bogdanPricope commented:
Hi Ola,

Nice catch. It is nice to see that ofp is still dragging attention.
I am not sure who can review and push this change as Nokia & Enea are not active. But ARM has a maintainer in this project, right?
Myself, I am working on a commercial version of ofp, named nfp: https://github.com/NetInoSoftware/nfp 

Best regards,
Bogdan

WonderfulVoid commented:
I talked to someone at Nokia, they said they would handle GitHub pull requests.

JereLeppanen commented on include/api/ofp_sysctl.h:
```diff
@@ -54,6 +54,7 @@
 /*
  * Helper definitions
  */
+#undef __CONCAT
 #define	__CONCAT(x,y)	x ## y
```
Isn't it possible to simply remove this define?

WonderfulVoid replied:
That might be possible but then you must trust that __CONCAT is actually defined in some included header file and has an appropriate definition. I don't think this is a good idea.
Perhaps OFP should use its own definition of __CONCAT, e.g. OFP_CONCAT.

JereLeppanen replied:
> Perhaps OFP should use its own definition of __CONCAT, e.g. OFP_CONCAT.

Yes, I think this is the best option here. It should probably be _OFP_CONCAT, i.e. start with an underline to indicate an internal definition, not meant to be used by an application.

WonderfulVoid replied:
Symbols starting with underscore are reserved. I think OFP_CONCAT is the better choice or possibly OFP__CONCAT.

JereLeppanen replied:
Regardless, leading underscore is used for this purpose in both OFP and ODP. (Even if this convention is not totally consistent in OFP.) OFP__CONCAT looks odd, since currently that prefix is used only with some OFP__*_T_DECLARED macros. OFP_CONCAT does not in any way convey that it's an internal definition.

WonderfulVoid replied:
I can change it to e.g. _OFP_CONCAT.

JereLeppanen commented on src/ofp_timer.c:
```diff
@@ -481,7 +481,8 @@ int ofp_timer_cancel(odp_timer_t tim)
 			tmo = odp_timeout_from_event(timeout_event);
 			bufdata = odp_timeout_user_ptr(tmo);
-			odp_buffer_free(bufdata->buf);
+			if (bufdata != NULL)
+				odp_buffer_free(bufdata->buf);
```
Have you seen some problems with this? This user pointer should probably always be non-NULL, so if there's a problem somewhere, this condition just hides the problem, but it doesn't fix it.

WonderfulVoid replied:
Yes I had a crash here. I use my own ODP implementation but I can't see how that affects the issue, the timeout user pointer is under the control of the application?
I can debug this further and see the life cycle of the affected timeout.

WonderfulVoid replied:
This change is not needed. I had a bug in my clean room ODP implementation, the documentation for odp_timer_set_abs()  isn't very clear on how fields are copied from a timer to its timeout event.

bogdanPricope commented:
Welcome back, @JereLeppanen !

JereLeppanen commented:
Heh, thanks Bogdan. No comment on whether I really ever left, or came back.

JereLeppanen commented:
CI is now using github actions, so if you update this PR, please remember to rebase so that we get a CI run.

WonderfulVoid commented:
I updated one commit (088351e) and removed another (the NULL check).
Should I "Update with merge commit" (which seems to be default) or "Update with rebase" (which feels more natural to me)?

JereLeppanen commented:
> Should I "Update with merge commit" (which seems to be default) or "Update with rebase" (which feels more natural to me)?

Rebase, please.

JereLeppanen commented:
Any progress here? I think all that is needed is the rebase and also the commits need descriptions and signed-off-by tags.

WonderfulVoid commented:
Sorry for the delay, I was in Powerpoint mode for a month...
I rebased and updated the OFP_CONCAT commit per your request.
What kind of descriptions do you want for the commits? And who can sign off? I am all alone here with OFP work...

JereLeppanen commented:
Your signed-off-by tag is enough. Please also add

`Reviewed-by: Jere Leppänen <jere.leppanen@nokia.com>`

Description for each commit should state what is changed and why. Just one or two sentences is enough for these commits.

I missed this earlier, but the subject lines of the first and last commits should be prefixed with "api:", i.e.

```
api: Add OFP_ prefix to _IN_PORT_T_DECLARED symbol
api: Add OFP_ prefix to SOMAXCONN symbol
```

And finally, if you like, you could fix the checkpatch warnings: https://github.com/OpenFastPath/ofp/runs/5738785370?check_suite_focus=true You didn't touch those parts of the changed lines, so this is optional.
